### DISCUSSION TITLE
Error: spawn sh ENOENT when using "vercel build" with --cwd flag in npm monorepo

### DISCUSSION THREAD

**Discussion by danawoodman** [vercel/vercel#10202] [Help]
### Problem Description

I'm wondering if someone can point out what I'm doing wrong here but I'm attempting to make a custom deploy script in my npm monorepo that I can use to deploy any Vercel apps in my `apps/` dir. 

This is a test shell scrip I've been using to debug the issue:

```shell
#!/bin/zsh

DIR="apps/google-wallet"
export VERCEL_ORG_ID="..."
export VERCEL_TOKEN="..."
export VERCEL_PROJECT_ID="..."

npm install --global vercel@latest

vercel link -y --cwd $DIR
vercel pull -y --environment production $DIR
vercel build -y --prod --cwd $DIR
vercel deploy -y --prod --prebuilt $DIR
```

All the commands work up until `build` where I'm getting `Error: spawn sh ENOENT`.

I've also instead `cd`ed into the app directory and it still fails with the `EOENT` error.

My `vercel.json` (in the root of my repo) is: 

```json
{
	"$schema": "https://turbo.build/schema.json",
	"buildCommand": "VERCEL=true npx turbo run build",
	"github": { "silent": true, "enabled": false },
	"trailingSlash": false,
	"scope": "myorghere"
}
```

If I run build form the root of the repo instead of using `--cwd` it builds the first app in my `apps/` directory not the one I've linked. 

I've tried a lot of variations of the above, but it seems reliable that build fails when trying to specify a specific app either by `cd`ing or usign `--cwd`.

Vercel CLI version: 31.0.2

**Comment by danawoodman** [comment]
Another issue I'm seeing is if you omit the `--cwd` option it always defaults to the first project in your list of apps when running `vercel pull`, even if you specify a project in your `vercel link` command before. There is no option for specifying the project in `vercel pull`

If I `cd` into the directory I get the same `spawn sh ENOENT` and the debug path to the app is strangely: 

```
/apps/app-b/apps/app-a
```

It's appending `apps/app-a` on even tho I never specific that app anywhere, Vercel just seems to decide that I want to deploy app-a on its own

**Comment by janvorwerk** [comment]
Hello

I am facing the same issue and I am about to open an issue.
Here is what I did (very easy to reprdoduce)

    pnpm dlx create-turbo@latest

Then I upload the created repo to GitHub and deploy the two projects to my Vercel account:
- my-turborepo-web
- my-turborepo-docs

On my dev box, go to `my-turborepo` and run:

    env VERCEL_ORG_ID=... VERCEL_PROJECT_ID=... vercel pull --cwd apps/docs --token=...

=> all right, it downloads the project.json

```
{
  "projectId": "...",
  "orgId": "...",
  "settings": {
    "createdAt": 1693418631143,
    "framework": "nextjs",
    "devCommand": null,
    "installCommand": null,
    "buildCommand": null,
    "outputDirectory": null,
    "rootDirectory": "apps/docs",
    "directoryListing": false,
    "nodeVersion": "18.x"
  }
}
```


But if I try to build, 

        env VERCEL_ORG_ID=... VERCEL_PROJECT_ID=... vercel build --cwd apps/docs --token=...

It says:

    Error: ENOENT: no such file or directory, open '.../my-turborepo/apps/docs/apps/docs/package.json

Observe the duplicate `apps/docs` in the path.

If I edit the JSON above to set `rootDirectory": "",` then in works like a charm!

It seems that the rootDirectory should not be concatenated to the `--cwd` arg

**Comment by janvorwerk** [comment]
It seems to me that the problem lies in the fact that we have two contatenations: [here](https://github.com/vercel/vercel/blob/fc1e13c09928c654410b373fc1775c2b63c6ef4a/packages/cli/src/commands/build/index.ts#L192) and [there](https://github.com/vercel/vercel/blob/fc1e13c09928c654410b373fc1775c2b63c6ef4a/packages/cli/src/commands/build/index.ts#L297) I would say.... But it's hard to tell how it should be fixed...

**Comment by janvorwerk** [comment]
Simple workaround in my GitHub action : 

    sed -i.bak '/rootDirectory/d' ./apps/docs/.vercel/project.json

**Comment by anthonyshew** [comment]
Have either of you given `--repo` a try?
https://vercel.com/docs/cli/link#repo-alpha
